internal/runtime/funcs/wait_group.go

```diff
@@ -0,0 +1,88 @@
+package funcs
+
+import (
+	"context"
+	"sync"
+
+	"github.com/nevalang/neva/internal/runtime"
+)
+
+type waitGroup struct{}
+
+func (g waitGroup) Create(io runtime.FuncIO, _ runtime.Msg) (func(ctx context.Context), error) {
+	addIn, err := io.In.Port("add")
+	if err != nil {
+		return nil, err
+	}
+
+	sigOut, err := io.Out.Port("sig")
+	if err != nil {
+		return nil, err
+	}
+
+	return g.Handle(addIn, sigOut), nil
+}
+
+func (g waitGroup) Handle(
+	addIn,
+	sigOut chan runtime.Msg,
+) func(ctx context.Context) {
+	return func(ctx context.Context) {
+		var (
+			mu    sync.RWMutex
+			ready bool  // ready when we have received a nonnegative count, protected by mu.
+			count int64 // internal count, protected by mu.
+		)
+
+		reset := func() {
+			mu.Lock()
```

Comment:
why not just native wait group?

Reply:
Let's continue both topics here

> I can rewrite the example to use Go's sync.WaitGroup internally, but know that WaitGroup panics if the count is less than 0. This makes it very difficult to use for the implementation in Neva because -1 can and often does reach the counter before N which will cause a panic. We will need a temporary Int64 to hold negative numbers to avoid panic. It is "much easier to leave the current implementation".

I might be wrong but to me "temporary Int64" feels like just having a variable. How is it "much easier to leave the current implementation"? However, I won't push. Any correct implementation is ok for me.

However, I hope you'll be ok if one day I'll try to reimplement it with `sync.WaitGroup` if I see that it's easier

Reply:
I would also add that

> WaitGroup panics when the count gets below 0, since we may have multiple calls happening concurrently, we need to protect against this case.

To me feels like we just need to read from `count` before we read from `sig/add/sub` (second port). I might be wrong tho

Reply:
There's no one implementation, I was trying to make something that would always accept a new int to count but I realize this might be very bad now because consts are sent in a for loop...? I'll stash this and rework it to make it more like your instance of count with signal import